client/src/margo_client.c

```diff
@@ -1211,3 +1212,118 @@ static void unifyfs_unlink_callback_rpc(hg_handle_t handle)
+
+    /* get handle to rpc function */
+    hg_handle_t handle = create_handle(client_rpc_context->rpcs.get_gfids_id);
+
+    /* fill in input struct */
+    unifyfs_get_gfids_in_t in;
+    in.app_id     = (int32_t) client->state.app_id;
+    in.client_id  = (int32_t) client->state.client_id;
+    /* TODO: What are app_id and client_id for?!? */
+
+    /* call rpc function */
+    LOGDBG("invoking the get_gfids rpc function in client");
+    double timeout = client_rpc_context->timeout;
+    ret = forward_to_server(handle, &in, timeout);
+    if (ret != UNIFYFS_SUCCESS) {
+        LOGERR("forward of get_gfids rpc to server failed");
+        goto exit;
+    }
+
+    /* decode response */
+    unifyfs_get_gfids_out_t out;
+    get_output_hret = margo_get_output(handle, &out);
+    if (get_output_hret != HG_SUCCESS) {
+        LOGERR("margo_get_output() failed - %s",
+               HG_Error_to_string(get_output_hret));
+        ret = UNIFYFS_ERROR_MARGO;
+        goto exit;
+    }
+
+    LOGDBG("Got response ret=%" PRIi32, out.ret);
+    ret = (int) out.ret;
+    if (ret != (int)UNIFYFS_SUCCESS) {
+        goto exit;
+    }
+
+    LOGDBG("Number of GFIDs returned: %d", out.num_gfids);
+    *num_gfids = out.num_gfids;
+
+    // Pull the bulk data (the list of gfids) over from the server
+    hg_bulk_t local_bulk;
+    hg_size_t buf_size = out.num_gfids * sizeof(**gfid_list);
+
+    // Allocate local memory for the list of GFIDs
+    int* _gfid_list = calloc(out.num_gfids, sizeof(**gfid_list));
+
+    // Figure out some margo-specific info that we need for the transfer
+    const struct hg_info* info = margo_get_info(handle);
+    // TODO: Is this the correct handle??
+    hg_addr_t server_addr = info->addr;
+    // address of the bulk data on the server side
+    margo_instance_id mid = margo_hg_handle_get_instance(handle);
+    // TODO: Is this the correct handle??
+
+    bulk_create_hret = margo_bulk_create(mid, 1, (void**)&_gfid_list, &buf_size,
+                                         HG_BULK_WRITE_ONLY, &local_bulk);
+    if (bulk_create_hret != HG_SUCCESS) {
+        LOGERR("margo_bulk_create() failed - %s",
+               HG_Error_to_string(bulk_create_hret));
+        ret = UNIFYFS_ERROR_MARGO;
+        goto exit;
+    }
+
+    hg_return_t hret = margo_bulk_transfer(mid, HG_BULK_PULL, server_addr,
+                                           out.bulk_gfids, 0, local_bulk,
+                                           0, buf_size);
+    if (hret != HG_SUCCESS) {
+        LOGERR("margo_bulk_transfer() failed - %s", HG_Error_to_string(hret));
+        ret = UNIFYFS_ERROR_MARGO;
+        goto exit;
+    }
+
+    // copy the pointer so it can be returned to the caller
+    // (Caller will need to eventually call free() on this pointer!)
+    *gfid_list = _gfid_list;
+
+exit:  // normally, we use "out", but that's also the name of a variable
+    /* free resources */
+    if (ret != UNIFYFS_SUCCESS) {
+        free(_gfid_list);
```

Comment:
Could this have an uninitialized value, which would then cause an error in free()?  For example, if an error is detected at line 1249?

Reply:
Yep.  Good catch.  I'll make sure `_gfid_list` gets initialized to NULL up at the top of the function.  (Calling free() on NULL is explicitly allowed.)